src/backend/access/nbtree/nbtsearch.c

```diff
@@ -1100,6 +1167,37 @@ _bt_first(IndexScanDesc scan, ScanDirection dir)
 		}
 	}
 
+	/* Neon: initialize prefetch */
+	so->n_prefetch_requests = 0;
+	so->n_prefetch_blocks = 0;
+	so->last_prefetch_index = 0;
+	so->next_parent = P_NONE;
+	so->prefetch_maximum = IsCatalogRelation(rel)
+		? effective_io_concurrency
+		: get_tablespace_io_concurrency(rel->rd_rel->reltablespace);
+
+	if (scan->xs_want_itup) /* index only scan */
+	{
+		if (enable_indexonlyscan_prefetch)
+		{
+			/* We disable prefetch for parallel index-only scan.
+			 * Neon prefetch is efficient only if prefetched blocks are accessed by the same worker
+			 * which issued prefetch request. The logic of splitting pages between parallel workers in
+			 * index scan doesn't allow to satisfy this requirement.
+			 * Also prefetch of leave pages will be useless if expected number of rows fits in one page.
+			 */
+			if (scan->parallel_scan || scan->rows_estimation <= MaxTIDsPerBTreePage)
```

Comment:
The reasons of disabling prefetch for parallel index-only scan and for low row count estimate are different.
Implementation of parallel index only scan makes prefetch almost useless, because fetching of next leaf page is performed  in critical section to synchronize parallel workers.  So parallel index only scan con not some how improve speed of loading B-Tree pages. This is why times of query execution with index scan are almost the same for parallel and non-parallel plans if index doesn't fir in shared buffers.

We can not do here something like in case of bitmap heap scan prefetch because there is no way to predict which leaf pages will be needed without actual traversing of the tree.

Concerning low rowcount estimates - I agree with you.  Te idea here was not to prefetch leave pages if we need just few tuples. But you are right - rowcount estimation is not the  best thing to look at, because number of filtered tuples may be much larger. So we actually need estimation for number of scanned tuples. But I do not know how to get it.
So my decision to use this criteria was was on the assumption that typically (for well defined index) number of filtered tuples should be small and rowcount estimation is almost the same as number of traversed rows.

Reply:
'Slow start' works well for index scan (when we fetch referenced heap tuples).
We really can start with prefetch_distance=0, so doesn't prefetch any heap tuple and then increase prefetch distance.
And this is how it is implemented now.

But for index-only scan situation is different: we need to prefetch leave pages of B-Tree and to be able to do it we need to read parent page. It is not a big deal because we in any case has traversed it and so it should be present in shared buffers,
but still I prefer not to do it for "point selects". And it seems to be the most common use case for index-only scans.
Yes, I agree with you that rownumber estimation may be not the best criteria for it. But I prefer too be more conservative here: better do not use prefetch when it may be applicable for filtering large number of records, rather then slowdown unique index lookups.

Reply:
So you suggest not to try to read parent pretch and so prefetch any other leave pages until we reach end of the first leaf page? May be it makes sense...

But do I correctly understand that you do not like idea of taken in account optimizer estimation of number of returned tuples for choosing prefetch strategy? Removing it certainly simplify this PR and reduce number of changes in Postgres core. But I still think that it is quit useful information and it  is already available so using it add no any  extra coast. May be it is used not in the most optimal way. But I think that it should be considered and may be not only for index-only scan.

Concerning visibility checks. I agree with you that it can be a big problem. Index-only scan is efficient only if most pages are marked as all-visible. If it is not true, then index-only scan can be slower than heap scan, especially in Neon.
Right now index-only scan is not trying to prefetch heap tuples. Just because it has no idea whether them will be needed or not. In principle it can use visibility map to check it. But I am not sure if should be done by prefetch. I think that it is better to do at optimizer level: it should choose index-only scan plan only if it is sure that most of heap pages are all visible.

Reply:
Ok, I have removed rows estimation criteria from criteria of applying prefetch for index-only scan.
But I preserved existed logic of prefetch of index-only scan:
we save all items from right-most parent in `_bt_scan` method because this page is in any cases traversed by `_bt_search` or `_bt_endpoint` and then start prefetching on leave pages incrementing prefetch distance in each call of `_bt_steppage`.

Reply:
Done